Rewrite Makefile to support unified build as part of husky subprojects

spanevin opened the pull request:
Corresponding PR to huskybse: https://github.com/huskyproject/huskybse/pull/11

spanevin committed:
Rewrite Makefile to support unified build as part of husky subprojects

spanevin committed:
Rewrite Makefile to support unified build as part of husky subprojects

spanevin committed:
Rewrite Makefile to support unified build as part of husky subprojects

spanevin committed:
Fix 'multiple definitions of...' linker error

dukelsky reviewed (changes requested):
The names of all four version strings in `version.h` must begin with the lowercase subproject name, followed by an underscore, like `bsopack_VER_BRANCH.`

When one runs bsopack and hpucode without options, they behave differently. While hpucode shows usage, bsopack says `Could not find Config-file.` It would be better if bsopack first shows usage and then the diagnostics `Could not find Config-file.`

Please check that the following is true.
 
If `-c` command line option is defined, then its value determines the `fidoconfig` location. If `-c` command line option is not defined but `FIDOCONFIG` environment variable is defined, then the environment variable determines the `fidoconfig` location. If neither `-c` command line option, nor `FIDOCONFIG` environment variable are defined, then the settings `CFGDIR` and `CFGNAME` defined in `huskymak.cfg` determine the `fidoconfig` location during compilation. If `CFGNAME` is not defined there, `config` is used as `CFGNAME.`

It would be good to include the above paragraph in the documentation.

Please replace README with README.md, written using Markdown language.

Please check all the docs (both headers and text) using the `LanguageTool`.

The authors of both `bsopack` and `hpucode` have left Fidonet long ago, so it is better instead of `Contacting the author` just give `The author` in the docs and do not provide their contacts, just the name.

I did not test the program work, since you will now support both `bsopack` and `hpucode`.

spanevin commented:
I'm not sure if I can support them as I know nothing about their code and I'm not good in c/c++ at all. I just want to use them, as my node based on 1.4 uses them, so I want them to be compiled if I want to upgrade with minimum effort. If they don't work as good as 1.4 (maybe they can break some message bases on x64 version, if nobody tested them), it's just waste of time... But we won't know if we can't compile them.
I'm not good in MAKE language as well, so I just copied makefiles from one of the most similar project and sed-ed them :) and verified that it works and installs all required files (binaries, mans, docs, infos).

I can try to update docs here based on your suggestions. But I don't know what the LanguageTool which you mentioned above is. And hpucode doesn't have any README so there's nothing to update, just VERSION maybe...

dukelsky commented:
https://languagetool.org/
You create a free account there, install their browser extension, and it will check what you write in a browser field.

spanevin committed:
Forgotten removal of a moved file source

spanevin committed:
Remove unneeded areafix references in Makefile

spanevin committed:
Rename version component constants to conform naming conventions of other submodules

spanevin commented:
> When one runs bsopack and hpucode without options, they behave differently. While hpucode shows usage, bsopack says `Could not find Config-file.` It would be better if bsopack first shows usage and then the diagnostics `Could not find Config-file.`
> 
> Please check that the following is true.

It looks like bsopack tries to read and process config files before it processes any cmdline option. -c is probably the only exception, and it just adds a value of this option to the config search path...
I have husky-1.4 installed on my system, with configs for it, so I can't reproduce the issue with "Could not find Config-file." error. In my case it finds the config and fails on parsing it, so I don't even have ability to get usage message at all.

I don't think I'm experienced enough to change/fix bsopack startup behavior, sorry. Maybe I'll work on it later, when I have more free time. Changing of existing behavior is not the goal of this PR.

spanevin committed:
Update documentation: remove contacts of authors as they've left Fidonet long time ago, fix spelling issues

spanevin committed:
Rewrite README with MarkDown language as README.md

spanevin committed:
Convert INSTALL to MarkDown as INSTALL.md

dukelsky commented:
> > When one runs bsopack and hpucode without options, they behave differently. While hpucode shows usage, bsopack says `Could not find Config-file.` It would be better if bsopack first shows usage and then the diagnostics `Could not find Config-file.`
> > Please check that the following is true.
> 
> It looks like bsopack tries to read and process config files before it processes any cmdline option. -c is probably the only exception, and it just adds a value of this option to the config search path... I have husky-1.4 installed on my system, with configs for it, so I can't reproduce the issue with "Could not find Config-file." error. In my case it finds the config and fails on parsing it, so I don't even have ability to get usage message at all.

So, you didn't build and run the app? OMG! The first thing you should do when you propose a pull request is building and running the changed application. So, please do it. If it doesn't work at all, and you are not going to fix it, there is no sense in the pull request. If you are going to fix it later, then it may be a subject of a pull request to follow this one.

> 
> I don't think I'm experienced enough to change/fix bsopack startup behavior, sorry. Maybe I'll work on it later, when I have more free time. Changing of existing behavior is not the goal of this PR.

OK. If you don't fix the `bsopack` startup behavior now, then it is better to exclude the paragraph about the behavior from the docs and include it after (and if) you fix it.

The content of the `INSTALL.md` has nothing to do with the proposed use of the Makefile!

There are excessive space characters at the end of the lines 75,76,79,84,85 of `README.md`. They are shown as “more green” in the diff. Actually, it is a small thing.

spanevin commented:
> > > When one runs bsopack and hpucode without options, they behave differently. While hpucode shows usage, bsopack says `Could not find Config-file.` It would be better if bsopack first shows usage and then the diagnostics `Could not find Config-file.`
> > > Please check that the following is true.
> > 
> > 
> > It looks like bsopack tries to read and process config files before it processes any cmdline option. -c is probably the only exception, and it just adds a value of this option to the config search path... I have husky-1.4 installed on my system, with configs for it, so I can't reproduce the issue with "Could not find Config-file." error. In my case it finds the config and fails on parsing it, so I don't even have ability to get usage message at all.
> 
> So, you didn't build and run the app? OMG! The first thing you should do when you propose a pull request is building and running the changed application.

I built it for sure. And ran. Build works. Run works as well, but fails on config parsing, and it's not surprising, as my config is for 1.4, not for 1.9, yet. I'm not going to spend time on converting configs until I have compiled versions of all required apps at least.

> OK. If you don't fix the `bsopack` startup behavior now, then it is better to exclude the paragraph about the behavior from the docs and include it after (and if) you fix it.

Ok, I'll remove it for now.

> The content of the `INSTALL.md` has nothing to do with the proposed use of the Makefile!

I just converted it to md. Do you want be to rollback?

> 
> There are excessive space characters at the end of the lines 75,76,79,84,85 of `README.md`. They are shown as “more green” in the diff. Actually, it is a small thing.

Are you familiar with MarkDown language? 2 spaces at the end of a line is a line break. Do you think I like it? No, I hate trailing spaces. But...

dukelsky commented:
> > The content of the `INSTALL.md` has nothing to do with the proposed use of the Makefile!
> 
> I just converted it to md. Do you want be to rollback?

The `INSTALL` file described the previous procedure. Now you want to change it. Thus, the description should be changed. You may refer to the `INSTALL.asciidoc` in `huskybse`.

spanevin commented:
> > > The content of the `INSTALL.md` has nothing to do with the proposed use of the Makefile!
> > 
> > 
> > I just converted it to md. Do you want be to rollback?
> 
> The `INSTALL` file described the previous procedure. Now you want to change it. Thus, the description should be changed. You may refer to the `INSTALL.asciidoc` in `huskybse`.

The first paragraph of the README.md:
```
# Instruction to build bsopack

Please read the files (especially README.Makefiles) in the huskybse package FIRST!
```
It refers to README.Makefiles which refers to INSTALL.asciidoc, so there is the reference to INSTALL.asciidoc, just indirect. And it was there even before rename. I just fixed the reference to huskybse project. Do you want to see direct reference to INSTALL.asciidoc instead of indirect reference through README.Makefiles?

The rest of the instruction was not correct even before my changes, since it didn't refer to ANY Makefile at all. Maybe it's possible to use the rest of the instruction with some minor modifications to build bsopack with legacy makefiles (there are some platform-specific makefiles in 'make' directory), but it is not the goal of this PR to check legacy build and fix it's description.

If you know better way to rephrase INSTALL.md taking unknown/untested things into account - please suggest.

dukelsky commented:
> > The content of the `INSTALL.md` has nothing to do with the proposed use of the Makefile!
> 
> I just converted it to md. Do you want be to rollback?

The `INSTALL` file described the previous procedure. Now you want to change it. Thus, the description should be changed. You may refer to the `INSTALL.asciidoc` in `huskybse`.


> > > > The content of the `INSTALL.md` has nothing to do with the proposed use of the Makefile!
> > > 
> > > 
> > > I just converted it to md. Do you want be to rollback?
> > 
> > 
> > The `INSTALL` file described the previous procedure. Now you want to change it. Thus, the description should be changed. You may refer to the `INSTALL.asciidoc` in `huskybse`.
> 
> The first paragraph of the README.md:
> 
> ```
> # Instruction to build bsopack
> 
> Please read the files (especially README.Makefiles) in the huskybse package FIRST!
> ```
> 
> It refers to README.Makefiles which refers to INSTALL.asciidoc, so there is the reference to INSTALL.asciidoc, just indirect. And it was there even before rename. I just fixed the reference to huskybse project. Do you want to see direct reference to INSTALL.asciidoc instead of indirect reference through README.Makefiles?
> 
> The rest of the instruction was not correct even before my changes, since it didn't refer to ANY Makefile at all. Maybe it's possible to use the rest of the instruction with some minor modifications to build bsopack with legacy makefiles (there are some platform-specific makefiles in 'make' directory), but it is not the goal of this PR to check legacy build and fix it's description.
> 
> If you know better way to rephrase INSTALL.md taking unknown/untested things into account - please suggest.

I admire your reluctance to do anything to bring the document into compliance with the PR.

spanevin committed:
Update INSTALL.md

spanevin committed:
Fix command line arguments processing, check args before reading config, fix config reading and clearing

spanevin commented:
> When one runs bsopack and hpucode without options, they behave differently. While hpucode shows usage, bsopack says `Could not find Config-file.` It would be better if bsopack first shows usage and then the diagnostics `Could not find Config-file.`

The difference is that hpucode requires at least one areamask to be specified in cmdline args, but all args of bsopack are optional.
I checked readConfig() code of fidoconfig package, it does not throw an exception but does exit() by itself if it can't find config file, so there is no way to catch this case in bsopack code and show usage info. The only way is to re-implement config search logic, but why do we need fidoconf package then? So it it not a solution.

> 
> Please check that the following is true.
> 
> If `-c` command line option is defined, then its value determines the `fidoconfig` location. If `-c` command line option is not defined but `FIDOCONFIG` environment variable is defined, then the environment variable determines the `fidoconfig` location. If neither `-c` command line option, nor `FIDOCONFIG` environment variable are defined, then the settings `CFGDIR` and `CFGNAME` defined in `huskymak.cfg` determine the `fidoconfig` location during compilation. If `CFGNAME` is not defined there, `config` is used as `CFGNAME.`

readConfig() from fidoconf accepts one argument, a path to config file. If it's null - it does some internal magic to search for config file, probably using the alg you mentioned above, based on compile-time constants.

> 
> It would be good to include the above paragraph in the documentation.

If you description of the alg based on compile-time constants is correct, I think we can keep this message in the docs now.

I checked bsopack code and found a lot of issues with cmdline args processing and config file search priority. I did some changes to fix these issues, and also to read cmdline args BEFORE config file is read. So, now it processes cmdline options first, and shows usage message in case of -h or --help, as well as in case of errors in arguments.
It still can't show usage message if config file is specified correctly but can't be read by fidoconf code, as explained above.

I still don't have configs valid for 1.9 version yet, so I can't verify if it processes links correctly, but it compiles, shows help, and it shows config processing errors at least. So it works somehow.

dukelsky commented on src/config.c:
```diff
@@ -10,95 +10,100 @@
+                exit(-1);
             }
+
+            fidoConfigFile = (char *)smalloc(strlen(argv[i]));
```
AFAIR `strlen()` does not count the trailing `NUL`. So, it should be
`fidoConfigFile = (char *)smalloc(strlen(argv[i]) + 1);`

spanevin replied:
Good catch, I'll fix it.

spanevin replied:
done

dukelsky commented on src/config.c:
```diff
@@ -10,95 +10,100 @@
             }
+
+            fidoConfigFile = (char *)smalloc(strlen(argv[i]));
+            sprintf(fidoConfigFile, "%s", argv[i]);
```
Here, only `strlen(argv[i])` characters is printed and `fidoConfigFile[strlen(argv[i]) + 1]` is left undefined. We can use `snprintf`:
`snprintf(fidoConfigFile, strlen(argv[i]) + 1, "%s", argv[i]);`

spanevin replied:
Hmm, it was old code, not mine... Ok, I'll fix it, thanks.

spanevin replied:
done

spanevin committed:
Fix memory allocation and string copying on fidoconfig parameter reading
